File: jdroid-android-sample/src/main/java/com/jdroid/android/sample/ui/loading/BlockingLoadingFragment.java
package com.jdroid.android.sample.ui.loading;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import com.jdroid.android.androidx.lifecycle.Resource;
import com.jdroid.android.androidx.lifecycle.ResourceObserver;
import com.jdroid.android.fragment.AbstractFragment;
import com.jdroid.android.sample.R;
import com.jdroid.android.sample.androidx.SampleRepository;
import com.jdroid.android.sample.androidx.SampleViewModel;
import com.jdroid.android.sample.database.room.SampleEntity;

import androidx.annotation.Nullable;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

public class BlockingLoadingFragment extends AbstractFragment {

	private SampleViewModel sampleViewModel;
	private Observer<Resource<SampleEntity>> observer;

	@Override
	public Integer getContentFragmentLayout() {
		return R.layout.blocking_loading_fragment;
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);

		observer = new ResourceObserver<SampleEntity>() {

			@Override
			protected void onDataChanged(SampleEntity data) {
			}

			@Override
			protected void onStartLoading(@Nullable SampleEntity data) {
				getActivityIf().showLoading();
			}

			@Override
			protected void onStopLoading(@Nullable SampleEntity data) {
				getActivityIf().dismissLoading();
			}

			@Override
			protected AbstractFragment getFragment() {
				return BlockingLoadingFragment.this;
			}
		};
		sampleViewModel = ViewModelProviders.of(this).get(SampleViewModel.class);
		execute();
	}

	@Override
	public void onViewCreated(View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);

		Button fail = findView(R.id.fail);
		fail.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				execute();
			}
		});
	}

	private void execute() {
		if (sampleViewModel.getSampleEntity() != null) {
			sampleViewModel.getSampleEntity().removeObserver(observer);
		}
		sampleViewModel.setFailLoadFromNetwork(true);
		sampleViewModel.setLoadFromNetworkDelaySeconds(5);
		sampleViewModel.load(SampleRepository.ID, true).observe(this, observer);
	}

}

File: jdroid-android-sample/src/main/java/com/jdroid/android/sample/ui/loading/BlockingLoadingFragment.kt
package com.jdroid.android.sample.ui.loading

import android.os.Bundle
import android.view.View
import android.widget.Button
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.jdroid.android.androidx.lifecycle.Resource
import com.jdroid.android.androidx.lifecycle.ResourceObserver
import com.jdroid.android.fragment.AbstractFragment
import com.jdroid.android.sample.R
import com.jdroid.android.sample.androidx.SampleRepository
import com.jdroid.android.sample.androidx.SampleViewModel
import com.jdroid.android.sample.database.room.SampleEntity

open class BlockingLoadingFragment : AbstractFragment() {

    private lateinit var sampleViewModel: SampleViewModel
    private lateinit var observer: Observer<Resource<SampleEntity>>

    override fun getContentFragmentLayout(): Int? {
        return R.layout.blocking_loading_fragment
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        observer = object : ResourceObserver<SampleEntity>() {

            override fun onDataChanged(data: SampleEntity) {}

            override fun onStartLoading(data: SampleEntity?) {
                getActivityIf()!!.showLoading()
            }

            override fun onStopLoading(data: SampleEntity?) {
                getActivityIf()!!.dismissLoading()
            }

            override fun getFragment(): AbstractFragment {
                return this@BlockingLoadingFragment
            }
        }
        sampleViewModel = ViewModelProviders.of(this).get(SampleViewModel::class.java)
        execute()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val fail = findView<Button>(R.id.fail)
        fail.setOnClickListener { execute() }
    }

    private fun execute() {
        sampleViewModel.sampleEntity?.removeObserver(observer)
        sampleViewModel.setFailLoadFromNetwork(true)
        sampleViewModel.setLoadFromNetworkDelaySeconds(5)
        sampleViewModel.load(SampleRepository.ID, true).observe(this, observer)
    }
}
